SIPS/sip-6.md

```diff
@@ -0,0 +1,184 @@
+
+The entropy can be accessed from within a Snap, using the `snap_getEntropy` JSON-RPC method.
+
+## Motivation
+
+Before this SIP, Snaps did not have a way to get some kind of deterministic entropy, without storing it on the disk. If
+the user deletes the Snap, or deletes the data of MetaMask, this entropy is lost. This SIP proposes a new way to get
+deterministic entropy, which is tied to the secret recovery phrase of the user, combined with the Snap ID. In this case,
+the entropy can always be re-created, as long as the user has a copy of their secret recovery phrase.
+
+## Specification
+
+> Formal specifications are written in TypeScript.
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",
+"NOT RECOMMENDED", "MAY", and "OPTIONAL" written in uppercase in this document are to be interpreted as described in
+[RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Deriving the entropy
+
+Entropy is derived using a
+[BIP 32](https://github.com/bitcoin/bips/blob/6545b81022212a9f1c814f6ce1673e84bc02c910/bip-0032.mediawiki) derivation
+path. This derivation path MUST start with the magic value `0xd36e6170` (`1399742832'` in BIP 32 notation). The
+following eight indices are based on a hash of the Snap ID, with an optional salt. The hashing algorithm of choice is
+`keccak256`.
+
+The hash of the Snap ID is calculated as follows:
+
+```typescript
+const hash = keccak256(snapId + keccak256(salt));
+```
+
+If the salt is not provided, an empty string MUST be used instead:
+
+```typescript
+const hash = keccak256(snapId + keccak256(''));
+```
+
+The hash is then split into eight big endian `uint32 | 0x80000000` integers. The resulting derivation path is a
+combination of the magic value, and the eight integers:
+
+```typescript
+const computedDerivationPath = getUin32Array(hash).map((index) => (index | 0x80000000) >>> 0);
+const derivationPath = [0xd36e6170, ...computedDerivationPath];
+```
+
+The entropy is then derived using the secret recovery phrase of the user, and the derivation path. The derivation
+algorithm of choice is `secp256k1`. The entropy is the private key of the derived key pair.
+
+```typescript
+const { privateKey: entropy } = bip32Derive(secretRecoveryPhrase, derivationPath);
+```
+
+`bip32Derive` is defined as the `CKDpriv` function in
+[BIP 32](https://github.com/bitcoin/bips/blob/6545b81022212a9f1c814f6ce1673e84bc02c910/bip-0032.mediawiki), using a root
+node created as per
+[BIP 39](https://github.com/bitcoin/bips/blob/6545b81022212a9f1c814f6ce1673e84bc02c910/bip-0039.mediawiki).
+
+### `snap_getEntropy` JSON-RPC method
+
+The `snap_getEntropy` JSON-RPC method is used to get the entropy for a Snap. It takes a single parameter, which is an
+object with the following properties:
+
+- `salt` (optional, `string`): A salt to use when deriving the entropy. If provided, this MUST be interpreted as a UTF-8
```

Comment:
How would a snap use `snap_getEntropy` for different chains? Would they need different salts or should we have a built-in coinType or similar param?

Reply:
`snap_getEntropy` can be used to get general entropy, so it's not specific to BIP-32/44 key derivation. Snaps wanting to derive keys for different chains can use the entropy as seed, and derive BIP-44 nodes from it.